### DISCUSSION TITLE
V2 API 반환값 통일하기

### DISCUSSION THREAD

**Discussion by scarf005** [jiphyeonjeon-42/backend#774] [Ideas]
### @scarf005 

- [x] bookmetas
- [x] books

### @Jiwon-Woo 

- [x] reviews

### @honeyl3ee 

- [x] likes

### @nyj001012 

- [x] users

### 고난이도, 후순위

- [ ] ~~lendings~~
- [ ] ~~reservations~~
- [ ] ~~tags~~

예시: pubdate -> publishedAt

1개, 1개 쿼리(나), N개 쿼리

**Comment by scarf005** [comment]
# reservation

`/books/{id}?lendings=&reservations=`

`/books/{id}/reservations`

`/reservations/{id}`

`/reservations/{book-meta}`

**Comment by scarf005** [comment]
## Stock (예시)
### /api/stock/search 

- /api/books/search 에 통합

**Comment by scarf005** [comment]
## BookMetas

- bookInfoId, bookinfo 필드와 겹침 -> 

```json
{
"id": 2,
  "title": "TCP IP 윈도우 소켓 프로그래밍",
  "author": "김선우",
  "publisher": "한빛아카데미",
  "isbn": "9788998756444",
  "image": "https://image.kyobobook.co.kr/images/book/xlarge/444/x9788998756444.jpg",
  "categoryId": 2,
  "category": "네트워크",
"pubishedAt": "2022-07-09-22:49:33",
  "updatedAt": "2022-07-09-22:49:33"
}
```
```json
	{
    		"id": 3,
    		"donator": null,
        "callSign": "C5.13.v1.c2",
    		"status": "ok",
    		"updatedAt": "2022-07-09-22:49:33"
	}
```

`/api/books/{id}`

개별 도서(book)랑 도서 정보(bookinfo)이 정보가 지나치게 겹침

```json
{
	"bookmeta": {
      "id": 2,
      "title": "TCP IP 윈도우 소켓 프로그래밍",
      "author": "김선우",
      "publisher": "한빛아카데미",
      "isbn": "9788998756444",
      "image": "https://image.kyobobook.co.kr/images/book/xlarge/444/x9788998756444.jpg",
      "categoryId": 2,
      "category": "네트워크",
	  "pubishedAt": "2022-07-09-22:49:33",
      "updatedAt": "2022-07-09-22:49:33"
	},
	"book": {
		"id": 3,
		"donator": null,
		"callSign": "C5.13.v1.c2",
		"status": "ok",
		"updatedAt": "2022-07-09-22:49:33"
	}
}
```

**Comment by scarf005** [comment]
## 카테고리: 어디에서는 id, 어디에서는 문자열

`createdAt`: `Date`
`pubdate`: "YYYYMMDD" 형식 string -> `pubishedAt`로 변경

**Comment by scarf005** [comment]
## 정리 가이드

### 플랫 경로를 지향하기 위해 `{id}`만으로 조회 가능하게 하자
`/books/info/{bookInfoId}/like` -> `/likes/{id}`

### 여러 정보를 가져온다면 중첩 형태를 쓰자

#### ko
```json5
{
	bookId: 1
	bookMetaId: 2
	categoryId: 3
	category: "과학"
	// book 필드...
	// bookMeta 필드...
}
```

#### ok
```json5
{
	// book 스키마 형태 유지
	book: {
		id: 1
		// book 필드...
	},
	book-meta: {
		id: 2
		// bookMeta 필드...
	},
	category: {
		id: 3
		name: "과학"
	},
}
```


# 수정: ID 기반 쿼리는 지나치게 복잡하여 처음 아이디어처럼 중첩 반환을 사용해주시면 될 것 같습니다

**Comment by Jiwon-Woo** [comment]
## Reviews
**1. GET `/api/reviews/my-reviews` -> `/api/v2/user/reviews`**

<br>

**2. GET `/api/book-info/{bookInfoId}/reviews` 삭제, GET `/api/v2/reviews` 에 통합**
  - query에 bookInfoId(bookMetaId) 추가

<br>

**3. PUT `/api/v2/reviews/{reviewsId}` -> PATCH `/api/v2/user/reviews`**
  - body에 disabled만 받을 수 있게

    ```json
    body
    {
      "visible": true
    }
    ```

<br>

**4. PATCH와 DELETE에서 사용하는 `reviewsId` (복수)를 `reviewId` (단수) or `id` 로 변경**

<br>

**5. POST `/api/v2/reviews` 에서 query 말고 body에 bookInfoId(bookMetaId) 넣기**
```json
body
{
  "bookInfoId": 1,
  "content": "리뷰 내용"
}
```

<br>

**6. GET `/api/v2/reviews`**

- nickname은 react query를 이용해서 캐싱할 예정이므로 userId만 보냄

```json
{
  "items": [
    {
      "id": 0,
      "userId": 1,
      "bookInfoId": 0,
      "createdAt": "2023-09-16T02:25:55.974Z",
      "title": "string",
      "content": "string",
      "visible": boolean
    }
  ],
  "meta": {
    "totalItems": 42,
    "totalPages": 5
  }
}
```

**Comment by Jiwon-Woo** [comment]
### 질문
- limit에서 perPage로 통일되나요?
- disabled에서 visibility로 통일되나요?

**Reply by scarf005** [reply]
visible: boolean 어떨까요?

**Comment by nyj001012** [comment]
## Users
### API 명세에서 타입 혼용
- `role`의 데이터 타입 string -> enum(literal union)
- `overDueDay`의 데이터 타입 string -> number
- `intraId`의 데이터 타입 string -> number

**Reply by scarf005** [reply]
role:  "user" | "cadet" | "librarian" | "staff" 식으로 string을 반환해도 괜찮을것 같아요
reservedNum: number로
penaltyEndDate: Date string으로

**Reply by nyj001012** [reply]
- `overDueDay`는 프론트엔드에서 처리

**Comment by honeyl3ee** [comment]
## Likes

### 플랫 경로 사용
- POST '/books/info/{bookInfoId}/like' -> /likes/{id}'
- GET '/books/info/{bookInfoId}/like' -> /likes/{id}'
- DELETE '/books/info/{bookInfoId}/like' -> /likes/{id}'

### 중첩 형태 사용
- POST '/books/info/{bookInfoId}/like'
```json5 
{
  "userId": 1,
  "bookInfoId": 0
} 
```

```json5
{
	user: {
		id: 1
	},
	book-meta: {
		id: 0
	},
}
```

- GET '/books/info/{bookInfoId}/like'
```json5 
{
  "bookInfoId": 0,
  "isLiked": true,
  "likeNum": 1
} 
```

```json5
{
	book-meta: {
		id: 0
	},
	isLiked: ture
	likeNum: 1
}
```

**Reply by scarf005** [reply]
[github의 starring api](https://docs.github.com/en/rest/activity/starring?apiVersion=2022-11-28)를 벤치마킹하면

POST `/user/likes/{book-metas-id}`
DELETE `/user/likes/{book-metas-id}`
GET `/user/likes`

문제점: 다른 모든 경로와 달리 likes경로만 id 경로가 다른 자원의 id임

**Reply by Jiwon-Woo** [reply]
POST /user/likes, body { "book-meta-id": 1 }
DELETE /user/likes/{like_id}
GET /user/likes
